src/test/java/com/google/devtools/build/lib/remote/BuildWithoutTheBytesIntegrationTest.java

```diff
@@ -1390,4 +1391,55 @@ public void remoteTreeFilesExpiredBetweenBuilds_buildRewound() throws Exception
     // Assert: target was successfully built
     assertValidOutputFile("a/bar.out", "file-inside\nupdated bar\n");
   }
+
+  @Test
+  public void sharedSubtreeLost_recoveredWithoutRewinding(@TestParameter boolean useDiskCache)
+      throws Exception {
+    // Arrange: Prepare workspace and populate both caches
+    if (useDiskCache) {
+      addOptions("--disk_cache=disk_cache_dir");
+    }
+    write(
+        "a/BUILD",
+        """
+        genrule(
+            name = "bar",
+            srcs = [
+                "dir",
+                "bar.in",
+            ],
+            outs = ["bar.out"],
+            cmd = "( cat $(location :dir)/foo.in; cat $(location :bar.in) ) > $@",
+        )
+
+        genrule(
+            name = "baz",
+            srcs = [
+                "dir",
+                "baz.in",
+            ],
+            outs = ["baz.out"],
+            cmd = "( cat $(location :dir)/foo.in; cat $(location :baz.in) ) > $@",
+        )
+        """);
+    write("a/dir/foo.in", "file-inside");
+    write("a/bar.in", "bar");
+    write("a/baz.in", "baz");
+
+    // This uploads the shared source directory tree to the remote cache.
+    buildTarget("//a:bar");
+    assertOutputDoesNotExist("a/bar.out");
+    getOutputBase().getRelative("action_cache").deleteTreesBelow();
+
```

Comment:
Can we also verify (either here or in a separate test case) that inputs aren't uploaded to the disk cache?